frontend/src/features/map_planning/store/RemoteActions.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  CreatePlantAction,
+  DeletePlantAction,
+  MovePlantAction,
+  TransformPlantAction,
+} from '../layers/plant/actions';
+import useMapStore from './MapStore';
+import { Action } from './MapStoreTypes';
+import {
+  CreatePlantActionDto,
+  DeletePlantActionDto,
+  MovePlantActionDto,
+  TransformPlantActionDto,
+} from '@/bindings/definitions';
+import { User } from 'oidc-client-ts';
+import z from 'zod';
+
+export function handleRemoteAction(ev: MessageEvent<unknown>, user: User) {
+  if (typeof ev.data !== 'string') {
+    console.error('Received non-string message from server');
+    return;
+  }
+
+  if (ev.data === 'connected') {
+    // Ignore the initial connected message
+    return;
+  }
+
+  let remoteAction: RemoteAction;
+  try {
+    const remoteActionJson = JSON.parse(ev.data);
+    remoteAction = RemoteActionSchema.parse(remoteActionJson);
+  } catch (e) {
+    console.error(e);
+    return;
+  }
+
+  if (remoteAction.userId === user.profile.sub) {
+    // Ignore actions from self, later this can be used to handle conflicts.
+    // see https://www.figma.com/blog/how-figmas-multiplayer-technology-works/#syncing-object-properties
+    return;
+  }
+
+  const action = convertToAction(remoteAction);
+
+  useMapStore.getState().__applyRemoteAction(action);
+}
+
+function convertToAction(remoteAction: RemoteAction): Action<unknown, unknown> {
+  switch (remoteAction.type) {
+    case 'CREATE_PLANT':
+      return new CreatePlantAction({ ...remoteAction.payload });
+    case 'DELETE_PLANT':
+      return new DeletePlantAction(remoteAction.payload.id);
+    case 'MOVE_PLANT':
+      return new MovePlantAction([{ ...remoteAction.payload }]);
+    case 'TRANSFORM_PLANT':
+      return new TransformPlantAction([{ ...remoteAction.payload }]);
+    default:
+      throw new Error(`Unknown remote action`) as never;
+  }
+}
+
+const RemoteCreatePlantActionSchema = z.object({
```

Comment:
What's the purpose of these zod schemas?

Reply:
They are type-safety related.
Because everything is coming from the same endpoint it was the simplest solution that came to my mind to add validation and type-safety at the same time.
We could probably also check for the type attribute of an action and "lie" about what it really contains by casting it.